gapis/api/vulkan/frame_loop.go

```diff
@@ -126,129 +139,237 @@ func newFrameLoop(ctx context.Context, c *capture.GraphicsCapture, numInitialCmd
 
-// TODO: Find out from which command are the start and the end the loop.
-func (f *frameLoop) getLoopStartAndEndCmd(ctx context.Context, Cmds []api.Cmd) (startCmd, endCmd api.Cmd) {
-	startCmd = Cmds[f.numInitialCmds]
-	endCmd = Cmds[len(Cmds)-1]
+	if f.loopTerminated == false {
+		if f.lastObservedCommand == api.CmdNoID {
+			log.W(ctx, "FrameLoop transform was applied to whole trace (Flush() has been called) without the loop starting.")
+		} else {
+			log.E(ctx, "FrameLoop: current frame is %v cmdId %v, cmd is %v.", f.frameNum, f.capturedLoopCmdIds[len(f.capturedLoopCmdIds)-1], f.capturedLoopCmds[len(f.capturedLoopCmds)-1])
+			log.F(ctx, true, "FrameLoop transform was applied to whole trace (Flush() has been called) mid loop. Cannot end transformation in this state.")
+		}
+	}
+}
 
-	return startCmd, endCmd
+func (f *frameLoop) PreLoop(ctx context.Context, out transform.Writer) {
+}
+func (f *frameLoop) PostLoop(ctx context.Context, out transform.Writer) {
 }
 
 func (f *frameLoop) detectChangedResource(ctx context.Context, startState *api.GlobalState) {
+
 	f.backupState = f.capture.NewUninitializedState(ctx)
 	f.backupState.Memory = startState.Memory.Clone()
-	for k, v := range startState.APIs {
-		s := v.Clone(f.backupState.Arena)
-		s.SetupInitialState(ctx)
-		f.backupState.APIs[k] = s
+
+	for apiState, graphicsApi := range startState.APIs {
+
+		clonedState := graphicsApi.Clone(f.backupState.Arena)
+		clonedState.SetupInitialState(ctx)
+
+		f.backupState.APIs[apiState] = clonedState
 	}
-	s := f.backupState
 
-	err := api.ForeachCmd(ctx, f.cmds[f.numInitialCmds:], func(ctx context.Context, id api.CmdID, cmd api.Cmd) error {
+	// Loop through each command mutating the shadow state and looking at what has been created/destroyed
+	err := api.ForeachCmd(ctx, f.capturedLoopCmds, func(ctx context.Context, cmdId api.CmdID, cmd api.Cmd) error {
+
+		cmd.Extras().Observations().ApplyWrites(f.backupState.Memory.ApplicationPool())
+
 		switch cmd.(type) {
+
 		// Buffers.
 		case *VkCreateBuffer:
 			vkCmd := cmd.(*VkCreateBuffer)
-			vkCmd.Extras().Observations().ApplyWrites(s.Memory.ApplicationPool())
-			buffer := vkCmd.PBuffer().MustRead(ctx, vkCmd, s, nil)
+			buffer := vkCmd.PBuffer().MustRead(ctx, vkCmd, f.backupState, nil)
+			log.D(ctx, "Buffer %v created.", buffer)
 			f.bufferCreated[buffer] = true
-			cmd.Mutate(ctx, id, f.backupState, nil, f.watcher)
+
 		case *VkDestroyBuffer:
 			vkCmd := cmd.(*VkDestroyBuffer)
-			vkCmd.Extras().Observations().ApplyReads(s.Memory.ApplicationPool())
 			buffer := vkCmd.Buffer()
 			log.D(ctx, "Buffer %v destroyed.", buffer)
 			f.bufferDestroyed[buffer] = true
 
 		// Images
 		case *VkCreateImage:
 			vkCmd := cmd.(*VkCreateImage)
-			vkCmd.Extras().Observations().ApplyWrites(s.Memory.ApplicationPool())
-			img := vkCmd.PImage().MustRead(ctx, vkCmd, s, nil)
+			img := vkCmd.PImage().MustRead(ctx, vkCmd, f.backupState, nil)
+			log.D(ctx, "Image %v created", img)
```

Comment:
We probably don't want to log everything here

Reply:
See above

Reply:
See above!